blue-xdma/src/XDMATestbench.bsv

```diff
@@ -0,0 +1,147 @@
+import FIFOF::*;
+import Clocks :: *;
+import GetPut :: *;
+import BlueAXI :: *;
+import BlueLib :: *;
+import AXI4LiteMaster :: *;
+import AXI4Slave :: *;
+import XDMADescriptorGenerator :: *;
+import StmtFSM :: *;
+import Counter :: *;
+import Config :: *;
+
+interface IfcTop;
+    (* prefix = "" *) interface IfcXDMADescriptorGeneratorFab fab;
+endinterface
+
+(* synthesize, clock_prefix = "axi_aclk", reset_prefix = "axi_aresetn" *)
+module mkXDMATestbench(IfcTop);
+    let c2hTestDescriptor = XDMADescriptor {
+        length: fromInteger(valueOf(TESTLENGTH)),
+        srcAddr: fromInteger(valueOf(TESTSRCADDR)),
+        dstAddr: fromInteger(valueOf(TESTDSTADDR))
+    };
+    let h2cTestDescriptor = XDMADescriptor {
+        length: fromInteger(valueOf(TESTLENGTH)),
+        srcAddr: fromInteger(valueOf(TESTDSTADDR)),
+        dstAddr: fromInteger(valueOf(TESTSRCADDR))
+    };
+    let xdmaDescriptorGenerator <- mkXDMADescriptorGenerator;
+
+    Reg#(Maybe#(AXI4_Read_Rq#(AXI4_SLAVE_ADDRSz, AXI4_SLAVE_IDSz, AXI4_SLAVE_USRSz))) readReq <- mkReg(tagged Invalid);
+    Reg#(Maybe#(AXI4_Write_Rq_Addr#(AXI4_SLAVE_ADDRSz, AXI4_SLAVE_IDSz, AXI4_SLAVE_USRSz))) writeReq <- mkReg(tagged Invalid);
+
+    Reg#(Bool) c2hInitiated <- mkReg(False);
+    Reg#(Bool) h2cInitiated <- mkReg(False);
+    Reg#(Bool) c2hFinished <- mkReg(False);
+    Reg#(Bool) h2cFinished <- mkReg(False);
+    Reg#(Bool) h2cCheckedFailed <- mkReg(False);
+
+    // The function generates the bit vector by concatenating 8 bytes,
+    // each of which is obtained by adding the input address to an offset and masking the result with 'hFF.
+    // This code is designed to be compatible with existing example tests.
+    // ex: 0x0706050403020100 (addr = 0)
+    function Bit#(AXI4_SLAVE_DATASz) generatePattern(Bit#(AXI4_SLAVE_ADDRSz) addr);
```

Comment:
use random stimuli, instead of fixed pattern

Reply:
The generated pattern here will change with the variation of the address, so it is not fixed. 

This is done to be compatible with the other test cases in the example project. If a random algorithm is used instead, it is highly likely that some tests from the example design will not be reusable later.